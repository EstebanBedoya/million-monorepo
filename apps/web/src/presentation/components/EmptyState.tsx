'use client';

interface EmptyStateProps {
  title?: string;
  message?: string;
  onReset?: () => void;
}

export function EmptyState({ 
  title = "No Properties Found",
  message = "We couldn't find any properties matching your criteria. Try adjusting your filters.",
  onReset 
}: EmptyStateProps) {
  return (
    <div className="flex flex-col items-center justify-center py-16 px-4 text-center" role="status">
      {/* Icon */}
      <div className="mb-6 p-6 rounded-full bg-secondary/10">
        <svg
          className="w-16 h-16 text-secondary"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          aria-hidden="true"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={1.5}
            d="M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6"
          />
        </svg>
      </div>

      {/* Content */}
      <h3 className="text-2xl font-semibold text-foreground mb-3">
        {title}
      </h3>
      <p className="text-secondary max-w-md mb-6">
        {message}
      </p>

      {/* Reset button */}
      {onReset && (
        <button
          onClick={onReset}
          className="btn-primary"
          aria-label="Clear all filters"
        >
          Clear All Filters
        </button>
      )}
    </div>
  );
}

