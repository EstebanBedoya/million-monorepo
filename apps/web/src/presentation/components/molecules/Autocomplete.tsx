'use client';

import { useState, useRef, useEffect } from 'react';
import { Search, ChevronDown, Loader2 } from 'lucide-react';
import { Input } from '@/presentation/components/atoms/Input';
import { cn } from '@/utils/cn';

export interface AutocompleteOption {
  value: string;
  label: string;
  description?: string;
}

interface AutocompleteProps {
  options: AutocompleteOption[];
  value: string;
  onChange: (value: string) => void;
  onSearch?: (query: string) => void;
  placeholder?: string;
  emptyMessage?: string;
  loading?: boolean;
  error?: boolean;
  disabled?: boolean;
  className?: string;
}

export function Autocomplete({
  options,
  value,
  onChange,
  onSearch,
  placeholder = 'Search...',
  emptyMessage = 'No results found',
  loading = false,
  error = false,
  disabled = false,
  className,
}: AutocompleteProps) {
  const [isOpen, setIsOpen] = useState(false);
  const [searchQuery, setSearchQuery] = useState('');
  const [highlightedIndex, setHighlightedIndex] = useState(-1);
  const containerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);

  const selectedOption = options.find((opt) => opt.value === value);
  const displayValue = selectedOption?.label || '';

  const filteredOptions = options.filter((option) =>
    option.label.toLowerCase().includes(searchQuery.toLowerCase())
  );

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (containerRef.current && !containerRef.current.contains(event.target as Node)) {
        setIsOpen(false);
        setSearchQuery('');
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  useEffect(() => {
    if (onSearch && searchQuery) {
      const timeoutId = setTimeout(() => {
        onSearch(searchQuery);
      }, 300);
      return () => clearTimeout(timeoutId);
    }
  }, [searchQuery, onSearch]);

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const query = e.target.value;
    setSearchQuery(query);
    setIsOpen(true);
    setHighlightedIndex(-1);
  };

  const handleOptionClick = (option: AutocompleteOption) => {
    onChange(option.value);
    setIsOpen(false);
    setSearchQuery('');
    setHighlightedIndex(-1);
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (!isOpen && (e.key === 'ArrowDown' || e.key === 'Enter')) {
      setIsOpen(true);
      return;
    }

    if (!isOpen) return;

    switch (e.key) {
      case 'ArrowDown':
        e.preventDefault();
        setHighlightedIndex((prev) =>
          prev < filteredOptions.length - 1 ? prev + 1 : prev
        );
        break;
      case 'ArrowUp':
        e.preventDefault();
        setHighlightedIndex((prev) => (prev > 0 ? prev - 1 : -1));
        break;
      case 'Enter':
        e.preventDefault();
        if (highlightedIndex >= 0 && filteredOptions[highlightedIndex]) {
          handleOptionClick(filteredOptions[highlightedIndex]);
        }
        break;
      case 'Escape':
        setIsOpen(false);
        setSearchQuery('');
        setHighlightedIndex(-1);
        break;
    }
  };

  const handleFocus = () => {
    setIsOpen(true);
  };

  return (
    <div ref={containerRef} className={cn('relative', className)}>
      <div className="relative">
        <Input
          ref={inputRef}
          type="text"
          value={isOpen ? searchQuery : displayValue}
          onChange={handleInputChange}
          onKeyDown={handleKeyDown}
          onFocus={handleFocus}
          placeholder={placeholder}
          disabled={disabled}
          error={error}
          icon={loading ? <Loader2 className="w-4 h-4 animate-spin" /> : <Search className="w-4 h-4" />}
          className="pr-10"
        />
        <button
          type="button"
          onClick={() => setIsOpen(!isOpen)}
          disabled={disabled}
          className="absolute right-3 top-1/2 -translate-y-1/2 text-secondary hover:text-foreground transition-colors disabled:opacity-50"
        >
          <ChevronDown
            className={cn(
              'w-4 h-4 transition-transform',
              isOpen && 'transform rotate-180'
            )}
          />
        </button>
      </div>

      {isOpen && (
        <div className="absolute z-50 w-full mt-1 bg-card border border-border rounded-md shadow-lg max-h-60 overflow-auto">
          {loading ? (
            <div className="p-4 text-center text-secondary">
              <Loader2 className="w-5 h-5 animate-spin mx-auto mb-2" />
              <p className="text-sm">Loading...</p>
            </div>
          ) : filteredOptions.length === 0 ? (
            <div className="p-4 text-center text-secondary text-sm">
              {emptyMessage}
            </div>
          ) : (
            <ul className="py-1">
              {filteredOptions.map((option, index) => (
                <li key={option.value}>
                  <button
                    type="button"
                    onClick={() => handleOptionClick(option)}
                    className={cn(
                      'w-full px-4 py-2 text-left hover:bg-accent/10 transition-colors',
                      highlightedIndex === index && 'bg-accent/10',
                      value === option.value && 'bg-accent/20 font-medium'
                    )}
                  >
                    <div className="text-sm text-foreground">{option.label}</div>
                    {option.description && (
                      <div className="text-xs text-secondary mt-0.5">
                        {option.description}
                      </div>
                    )}
                  </button>
                </li>
              ))}
            </ul>
          )}
        </div>
      )}
    </div>
  );
}

