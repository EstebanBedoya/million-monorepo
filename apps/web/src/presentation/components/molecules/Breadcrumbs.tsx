'use client';

import Link from 'next/link';
import { ChevronRight } from 'lucide-react';
import { Icon } from '../atoms/Icon';

export interface BreadcrumbItem {
  label: string;
  href?: string;
}

export interface BreadcrumbsProps {
  items: BreadcrumbItem[];
  className?: string;
}

export const Breadcrumbs = ({ items, className }: BreadcrumbsProps) => {
  return (
    <nav 
      aria-label="Breadcrumb" 
      className={className}
    >
      <ol className="flex items-center space-x-2 flex-wrap">
        {items.map((item, index) => {
          const isLast = index === items.length - 1;
          
          return (
            <li key={index} className="flex items-center">
              {index > 0 && (
                <Icon 
                  icon={ChevronRight} 
                  size="sm" 
                  className="text-secondary mx-2" 
                  aria-hidden={true}
                />
              )}
              
              {isLast || !item.href ? (
                <span 
                  className="text-sm font-medium text-foreground"
                  aria-current={isLast ? 'page' : undefined}
                >
                  {item.label}
                </span>
              ) : (
                <Link 
                  href={item.href}
                  className="text-sm text-secondary hover:text-accent transition-colors"
                >
                  {item.label}
                </Link>
              )}
            </li>
          );
        })}
      </ol>
    </nav>
  );
};

