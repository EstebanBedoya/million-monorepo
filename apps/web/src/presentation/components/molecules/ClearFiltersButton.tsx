'use client';

import { X } from 'lucide-react';
import { Button } from '../atoms/Button';
import { Icon } from '../atoms/Icon';

export interface ClearFiltersButtonProps {
  onClick: () => void;
  className?: string;
}

export const ClearFiltersButton = ({ onClick, className }: ClearFiltersButtonProps) => {
  return (
    <Button
      variant="secondary"
      onClick={onClick}
      className={`w-full lg:w-auto ${className}`}
      aria-label="Clear all filters"
    >
      <Icon icon={X} size="sm" />
      Clear All Filters
    </Button>
  );
};
