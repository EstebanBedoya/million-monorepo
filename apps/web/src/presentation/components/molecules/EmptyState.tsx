'use client';

import { Home } from 'lucide-react';
import { Button } from '../atoms/Button';
import { Icon } from '../atoms/Icon';

export interface EmptyStateProps {
  title?: string;
  message?: string;
  onReset?: () => void;
  className?: string;
}

export function EmptyState({ 
  title = "No Properties Found",
  message = "We couldn't find any properties matching your criteria. Try adjusting your filters.",
  onReset,
  className
}: EmptyStateProps) {
  return (
    <div className={`flex flex-col items-center justify-center py-16 px-4 text-center ${className}`} role="status">
      {/* Icon */}
      <div className="mb-6 p-6 rounded-full bg-secondary/10">
        <Icon 
          icon={Home} 
          size="xl" 
          className="text-secondary"
          aria-hidden={true}
        />
      </div>

      {/* Content */}
      <h3 className="text-2xl font-semibold text-foreground mb-3">
        {title}
      </h3>
      <p className="text-secondary max-w-md mb-6">
        {message}
      </p>

      {/* Reset button */}
      {onReset && (
        <Button
          onClick={onReset}
          aria-label="Clear all filters"
        >
          Clear All Filters
        </Button>
      )}
    </div>
  );
}
