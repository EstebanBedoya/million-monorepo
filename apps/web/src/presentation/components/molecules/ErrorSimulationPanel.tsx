'use client';

import { useState } from 'react';
import { Button } from '../atoms/Button';
import { Badge } from '../atoms/Badge';
import { Icon } from '../atoms/Icon';
import { Settings, Play, Pause, RotateCcw } from 'lucide-react';

export interface ErrorSimulationPanelProps {
  onSetErrorSimulation: (enabled: boolean, errorRate: number) => void;
  onSetErrorTypes: (types: ('timeout' | 'server' | 'network')[]) => void;
  className?: string;
}

export function ErrorSimulationPanel({
  onSetErrorSimulation,
  onSetErrorTypes,
  className = ''
}: ErrorSimulationPanelProps) {
  const [isEnabled, setIsEnabled] = useState(false);
  const [errorRate, setErrorRate] = useState(0.1);
  const [selectedErrorTypes, setSelectedErrorTypes] = useState<('timeout' | 'server' | 'network')[]>(['server']);

  const handleToggle = () => {
    const newEnabled = !isEnabled;
    setIsEnabled(newEnabled);
    onSetErrorSimulation(newEnabled, errorRate);
  };

  const handleErrorRateChange = (rate: number) => {
    setErrorRate(rate);
    if (isEnabled) {
      onSetErrorSimulation(true, rate);
    }
  };

  const handleErrorTypeToggle = (type: 'timeout' | 'server' | 'network') => {
    const newTypes = selectedErrorTypes.includes(type)
      ? selectedErrorTypes.filter(t => t !== type)
      : [...selectedErrorTypes, type];
    
    setSelectedErrorTypes(newTypes);
    onSetErrorTypes(newTypes);
  };

  const handleReset = () => {
    setIsEnabled(false);
    setErrorRate(0.1);
    setSelectedErrorTypes(['server']);
    onSetErrorSimulation(false, 0.1);
    onSetErrorTypes(['server']);
  };

  if (process.env.NODE_ENV !== 'development') {
    return null;
  }

  return (
    <div className={`fixed bottom-4 right-4 bg-card border border-border rounded-lg p-4 shadow-lg z-50 max-w-sm ${className}`}>
      <div className="flex items-center gap-2 mb-3">
        <Icon icon={Settings} size="sm" className="text-primary" />
        <h3 className="font-semibold text-foreground">Error Simulation</h3>
        <Badge variant={isEnabled ? 'destructive' : 'secondary'}>
          {isEnabled ? 'ON' : 'OFF'}
        </Badge>
      </div>

      {/* Toggle */}
      <div className="flex items-center gap-2 mb-3">
        <Button
          onClick={handleToggle}
          variant={isEnabled ? 'destructive' : 'default'}
          size="sm"
          className="gap-2"
        >
          <Icon icon={isEnabled ? Pause : Play} size="sm" />
          {isEnabled ? 'Disable' : 'Enable'}
        </Button>
        
        <Button
          onClick={handleReset}
          variant="outline"
          size="sm"
          className="gap-2"
        >
          <Icon icon={RotateCcw} size="sm" />
          Reset
        </Button>
      </div>

      {/* Error Rate */}
      <div className="mb-3">
        <label className="block text-sm font-medium text-foreground mb-1">
          Error Rate: {Math.round(errorRate * 100)}%
        </label>
        <input
          type="range"
          min="0"
          max="1"
          step="0.1"
          value={errorRate}
          onChange={(e) => handleErrorRateChange(parseFloat(e.target.value))}
          className="w-full"
          disabled={!isEnabled}
        />
      </div>

      {/* Error Types */}
      <div className="mb-3">
        <label className="block text-sm font-medium text-foreground mb-2">
          Error Types:
        </label>
        <div className="flex flex-wrap gap-2">
          {(['server', 'timeout', 'network'] as const).map((type) => (
            <Button
              key={type}
              onClick={() => handleErrorTypeToggle(type)}
              variant={selectedErrorTypes.includes(type) ? 'default' : 'outline'}
              size="sm"
              className="text-xs"
            >
              {type}
            </Button>
          ))}
        </div>
      </div>

      {/* Info */}
      <p className="text-xs text-secondary">
        This panel only appears in development mode.
      </p>
    </div>
  );
}
