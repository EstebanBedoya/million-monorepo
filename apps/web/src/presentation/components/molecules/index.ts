export { SearchInput } from './SearchInput';
export { PriceRange } from './PriceRange';
export { PriceSuggestions } from './PriceSuggestions';
export { PropertyTypeSelect } from './PropertyTypeSelect';
export { PropertyDetails } from './PropertyDetails';
export { LocationInfo } from './LocationInfo';
export { ClearFiltersButton } from './ClearFiltersButton';
export { PageButton } from './PageButton';
export { PageNumberButton } from './PageNumberButton';
export { EmptyState } from './EmptyState';
export { PropertyCardSkeleton, PropertyListSkeleton } from './PropertyCardSkeleton';
export { PaginationSkeleton } from './PaginationSkeleton';
export { ThemeToggle } from './ThemeToggle';
export { AsyncBoundary } from './AsyncBoundary';
export { ErrorState } from './ErrorState';
export { LoadingState } from './LoadingState';
export { Breadcrumbs } from './Breadcrumbs';
