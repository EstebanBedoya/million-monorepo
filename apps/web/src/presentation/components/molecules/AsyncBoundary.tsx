'use client';

import { ReactNode } from 'react';
import { PropertyListSkeleton } from './PropertyCardSkeleton';
import { EmptyState } from './EmptyState';
import { ErrorState } from './ErrorState';

export interface AsyncBoundaryProps {
  children: ReactNode;
  loading?: boolean;
  error?: Error | string | null;
  isEmpty?: boolean;
  emptyTitle?: string;
  emptyMessage?: string;
  onRetry?: () => void;
  onClearFilters?: () => void;
  skeletonCount?: number;
  className?: string;
}

export function AsyncBoundary({
  children,
  loading = false,
  error = null,
  isEmpty = false,
  emptyTitle,
  emptyMessage,
  onRetry,
  onClearFilters,
  skeletonCount = 12,
  className
}: AsyncBoundaryProps) {
  // Show loading state
  if (loading) {
    return <PropertyListSkeleton count={skeletonCount} className={className} />;
  }

  // Show error state
  if (error) {
    return (
      <ErrorState 
        error={error}
        onRetry={onRetry}
        className={className}
      />
    );
  }

  // Show empty state
  if (isEmpty) {
    return (
      <EmptyState 
        title={emptyTitle}
        message={emptyMessage}
        onReset={onClearFilters}
        className={className}
      />
    );
  }

  // Show content
  return <>{children}</>;
}
