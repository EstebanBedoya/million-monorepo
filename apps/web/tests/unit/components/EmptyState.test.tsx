import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { EmptyState } from '../../../src/presentation/components/molecules/EmptyState';

describe('EmptyState', () => {
  it('should render with default props', () => {
    render(<EmptyState />);

    expect(screen.getByText('No Properties Found')).toBeInTheDocument();
    expect(screen.getByText(/We couldn't find any properties matching your criteria/i)).toBeInTheDocument();
  });

  it('should render with custom title and message', () => {
    render(
      <EmptyState 
        title="Custom Title" 
        message="Custom message here"
      />
    );

    expect(screen.getByText('Custom Title')).toBeInTheDocument();
    expect(screen.getByText('Custom message here')).toBeInTheDocument();
  });

  it('should render reset button when onReset is provided', () => {
    const mockOnReset = jest.fn();
    render(<EmptyState onReset={mockOnReset} />);

    const resetButton = screen.getByRole('button', { name: /Clear all filters/i });
    expect(resetButton).toBeInTheDocument();
  });

  it('should not render reset button when onReset is not provided', () => {
    render(<EmptyState />);

    const resetButton = screen.queryByRole('button', { name: /Clear all filters/i });
    expect(resetButton).not.toBeInTheDocument();
  });

  it('should call onReset when button is clicked', async () => {
    const mockOnReset = jest.fn();
    render(<EmptyState onReset={mockOnReset} />);

    const resetButton = screen.getByRole('button', { name: /Clear all filters/i });
    await userEvent.click(resetButton);

    expect(mockOnReset).toHaveBeenCalledTimes(1);
  });

  it('should have proper accessibility attributes', () => {
    render(<EmptyState />);

    const container = screen.getByRole('status');
    expect(container).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(<EmptyState />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

