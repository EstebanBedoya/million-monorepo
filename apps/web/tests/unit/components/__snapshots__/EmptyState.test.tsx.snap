// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyState should match snapshot 1`] = `
<div
  class="flex flex-col items-center justify-center py-16 px-4 text-center"
  role="status"
>
  <div
    class="mb-6 p-6 rounded-full bg-secondary/10"
  >
    <svg
      aria-hidden="true"
      class="w-16 h-16 text-secondary"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
      />
    </svg>
  </div>
  <h3
    class="text-2xl font-semibold text-foreground mb-3"
  >
    No Properties Found
  </h3>
  <p
    class="text-secondary max-w-md mb-6"
  >
    We couldn't find any properties matching your criteria. Try adjusting your filters.
  </p>
</div>
`;
